### PR TITLE
Implement a lot of missing features for Wolf and Improve tameable

### PR DESCRIPTION
**Pull request by GoodLucky777**
Implement a lot of missing features for Wolf and Improve tameable

### CODE REVIEW

File: src/main/java/nukkitcoders/mobplugin/EventListener.java [hunk 1]
```diff
@@ -292,5 +363,31 @@ public void PlayerMoveEvent(PlayerMoveEvent ev) {
             }
         }
     }
+    
+    @EventHandler
+    public void EntityDamageByEntityEvent(EntityDamageByEntityEvent ev) {
```
Comment (PetteriM1) [inline src/main/java/nukkitcoders/mobplugin/EventListener.java:368]: I think we should create a config option to disable this feature. Also ignoring cancelled damage events would be good.

File: src/main/java/nukkitcoders/mobplugin/entities/monster/walking/Wolf.java [hunk 2]
```diff
@@ -174,4 +393,63 @@ public void setCollarColor(DyeColor color) {
         this.setDataProperty(new IntEntityData(DATA_COLOUR, color.getColor().getRGB()));
         this.collarColor = color;
     }
+    
+    @Override
+    public boolean canDespawn() {
+        return false;
```
[the remark this answers is not in the document]
Comment (GoodLucky777) [reply src/main/java/nukkitcoders/mobplugin/entities/monster/walking/Wolf.java:399]: In minecraft wiki it says untamed wolf doesn't despawn  https://minecraft.fandom.com/wiki/Wolf#Behavior_and_appearance
However I don't know it works in Bedrock Edition. Also in https://minecraft.fandom.com/wiki/Spawn#Despawning despawn include hostile wolves but there was no despawn exclude in Bedrock Edition.

[the remark this answers is not in the document]
Comment (GoodLucky777) [reply src/main/java/nukkitcoders/mobplugin/entities/monster/walking/Wolf.java:399]: @PetteriM1 so what should I put here?

[the remark this answers is not in the document]
Comment (PetteriM1) [reply src/main/java/nukkitcoders/mobplugin/entities/monster/walking/Wolf.java:399]: Our spawning/despawning rules don't quite match vanilla anyways. I would let them despawn if they have no owner.

**Comment by GoodLucky777** [comment]
Ready to be reviewed @PetteriM1

File: src/main/java/nukkitcoders/mobplugin/EventListener.java [hunk 3]
```diff
@@ -64,6 +85,57 @@ public void EntityDeathEvent(EntityDeathEvent ev) {
...
+            ((Tameable) baseEntity).getOwner().dataPacket(tameDeathMessage);
+        }
+    }
+    
+    private void handleAttackedEntityAngry(EntityCreature entity) {
+        if (!(entity.getLastDamageCause() instanceof EntityDamageByEntityEvent)) return;
+        
+        Entity damager = ((EntityDamageByEntityEvent) entity.getLastDamageCause()).getDamager();
+        if (damager instanceof Wolf || damager instanceof IronGolem) { // TODO: Improve Iron Golem check
+            ((Wolf) damager).isAngryTo = -1L;
```
Comment (PetteriM1) [inline src/main/java/nukkitcoders/mobplugin/EventListener.java:135]: I don't think that casting IronGolem to Wolf is a very good idea. IronGolem and SnowGolem should have their own check where only `isAngryTo` is set.

**Pull request merged**